uber/models/department.py

```diff
@@ -0,0 +1,389 @@
+
+    @property
+    def required_roles_ids(self):
+        return [str(r.id) for r in self.required_roles]
+
+    @required_roles_ids.setter
+    def required_roles_ids(self, value):
+        self._set_relation(DeptRole, 'required_roles', value)
+
+    @property
+    def hours(self):
+        hours = set()
+        for i in range(self.duration):
+            hours.add(self.start_time + timedelta(hours=i))
+        return hours
+
+    @property
+    def end_time(self):
+        return self.start_time + timedelta(hours=self.duration)
+
+    def no_overlap(self, attendee):
+        before = self.start_time - timedelta(hours=1)
+        after = self.start_time + timedelta(hours=self.duration)
+        return not self.hours.intersection(attendee.hours) and (
+            before not in attendee.hour_map
+            or not attendee.hour_map[before].extra15
+            or self.department_id == attendee.hour_map[before].department_id
+        ) and (
+            after not in attendee.hour_map
+            or not self.extra15
+            or self.department_id == attendee.hour_map[after].department_id
+        )
+
+    @property
+    def slots_taken(self):
+        return len(self.shifts)
+
+    @property
+    def slots_untaken(self):
+        return max(0, self.slots - self.slots_taken)
+
+    @property
+    def is_setup(self):
+        return self.start_time < c.EPOCH
+
+    @property
+    def is_teardown(self):
+        return self.start_time >= c.ESCHATON
+
+    @property
+    def real_duration(self):
+        return self.duration + (0.25 if self.extra15 else 0)
+
+    @property
+    def weighted_hours(self):
+        return self.weight * self.real_duration
+
+    @property
+    def total_hours(self):
+        return self.weighted_hours * self.slots
+
+    def _potential_volunteers(
+            self, staffing_only=False, order_by=Attendee.full_name):
+        """
+        Return a list of attendees who:
+
+            1. Are assigned to this job's department.
+            2. Are allowed to work this job (job has no required roles
+               or the attendee's department membership fulfills all the
+               required roles).
+
+        Args:
+            staffing_only: Restrict result to attendees where staffing==True.
+            order_by: Order by another Attendee attribute.
+        """
+        query = self.session.query(Attendee)
+
+        if staffing_only:
+            query = query.filter(Attendee.staffing == True)  # noqa: E712
+
+        if self.required_roles:
+            query = query.join(Attendee.dept_roles, aliased=True).filter(and_(
+                *[DeptRole.id == r.id for r in self.required_roles]))
+        else:
+            query = query.join(Attendee.dept_memberships, aliased=True).filter(
+                DeptMembership.department_id == self.department_id)
+
+        return query.order_by(order_by).all()
+
+    @property
+    def capable_volunteers_opts(self):
```

Comment:
Huh... I'm assuming this is repeating functionality that was already there, but why would we populate an options() dropdown with volunteers who could work the job but don't necessarily have the availability for it?

Reply:
Performance.  Calculating the availability is expensive, since it requires us to load every shift and do a bunch of calculations for every volunteer in the dropdown.  We found that this severely impacted the load times of certain commonly used pages on-site, so we figured it was okay to just show everyone vaguely eligible and display a sensible error message if you accidentally tried to assign someone to a shift at the same time as another shift they already had.